### ISSUE TITLE
Roadmap for anywidget

### ISSUE THREAD

**Issue by paddymul** [manzt/anywidget#102]
I really like what I see so far from anywidget.  A couple of questions that would probably be addressed by a roadmap:

1. Where do you see anywidget fitting into the ipywidgets ecosystem, alongside related projects like dash, bokeh, holoviz?
2. Does anywidget have a story for packaging your widget up into a python package?
3. Does anywidget have a story for typescript?
4. How can anywidget help with testing?  I would love a guide and examples for how to test the  js/ts of a widget without having to restart/reevaluate the widget cell each time I make a change.  Some examples of tests and explanations of why they work to avoid that workflow would get me out of the whole pattern.

What else are you planning to build into anywidget?

Keep up the good work, Paddy.

**Comment by juba** [comment]
I am not an anywidget developer or maintainer but regarding your second point, I just published an early version of a widget as a Python package  and it seems to be working nicely and the process was quite straightforward.

**Comment by manzt** [comment]
Thanks for the interest in the project! If you haven't had the chance, I'd recommend taking a look at the docs because I've tried to address many of these questions there!

> Where do you see anywidget fitting into the ipywidgets ecosystem, alongside related projects like dash, bokeh, holoviz?

From the [blog](https://anywidget.dev/blog/introducing-anywidget/#hello-anywidget):

>  "[anywidget] is not a new widgets framework, but rather an abstraction around Jupyter Widgets that leverages ESM."

**anywidget** is a minimal layer on top of classic Jupyter Widgets. Widgets created with anywidget work just like widgets created with `ipywidgets`, it's just makes it a _lot_ easier to get started. I'm not that familiar with dash/bokeh/holoviz, but I think it would be fair to say that those projects are higher-level abstractions that let you built web-based visualizations/user interfaces from Python.

> Does anywidget have a story for packaging your widget up into a python package?

Yes, as others have noted you just need to include `anywidget` as a dependency in your python package. This feature is why I created the library!

> Does anywidget have a story for typescript?

TL;DR - use `esbuild`. You can write TypeScript/JSX and transpile your code on the fly, as we describe in the [docs](https://anywidget.dev/en/bundling/#development).

In order to write TypeScript you need a JavaScript tooling and some type of build system. I have some ideas for integrating this type of tooling within anywidget, but it feels like feature creep and something that is already handled really well via existing tools. Part of the reason classic jupyter widgets are frustrating to development is that they require the use of webpack. I don't want to make users have the same requirement. They should be able to write code that runs directly in the browser without modification, and opt-in to complexity.

As an aside, I personally love using TypeScript via [JSDoc](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html) in anywidget. You can write pure ESM that requires no compilation and still statically check the files with the TypeScript complier. It's the best of both worlds IMO.

> How can anywidget help with testing? I would love a guide and examples for how to test the js/ts of a widget without having to restart/reevaluate the widget cell each time I make a change. Some examples of tests and explanations of why they work to avoid that workflow would get me out of the whole pattern.

I'm not sure I understand what you mean. Are you asking about unit tests for your JavaScript, or seeing code changes during development. I think anywidget helps a lot in both cases:

- Unit tests: your widget front-end code must be ESM. ESM works out of the box with tools like [vitest](https://vitest.dev/guide/), and you could just import your module and run tests

```javascript
import { test, expect } from "vitest";
import { render } from "./widget.js"; // your widget

const mockView = {}; // would need to mock what Jupyter Widgets provides

test("it renders", async () => {
   render(mockView);
})
```

- Live development: Have you seen anywidget's [hot-module replacement](https://anywidget.dev/blog/anywidget-02/#native-hot-module-replacement-hmr)? It brings a modern dev experience to widget developers. The link also includes a video where I demo this feature. No need to reload the page or code to see changes. It's automatic.


> What else are you planning to build into anywidget?

I'm a PhD student and develop a lot of visualization tools. I created anywidget because I wanted to make it easier to bring these tools to Jupyter notebooks and Python. As maintainer of several regular Jupyter Widgets the maintaince burden was too extreme, and anywidget solves this problem for me (and others)!
